File: sources/ShmupWarz/Controllers/GameController.fs
namespace ShmupWarz
open System
open Bosco.ECS
open UnityEngine

type GameController () =
    inherit MonoBehaviour ()

    let createSystems(world:World) =
        world.Add(new MovementSystem(world))
        world.Add(new PlayerInputSystem(world))
        //world.Add(new SoundEffectSystem(world))
        world.Add(new CollisionSystem(world))
        world.Add(new ExpiringSystem(world))
        world.Add(new EntitySpawningTimerSystem(world))
        //world.Add(new ParallaxStarRepeatingSystem(world))
        //world.Add(new ColorAnimationSystem(world))
        world.Add(new ScaleAnimationSystem(world))
        world.Add(new RemoveOffscreenShipsSystem(world))
        world.Add(new RenderPositionSystem(world))
        world.Add(new ViewManagerSystem(world))

    let world = World.Create(Component.TotalComponents)

    (** *)
    member this.Start () =
        createSystems(world)
        world.Initialize()
        world.CreatePlayer()

    (** *)
    member this.Update () =
        world.Execute()

File: sources/src/controllers/GameController.ts
module schmupwarz {
  "use strict";

  import Pools = schmupwarz.Pools;
  import Systems = entitas.Systems;
  import Entity = entitas.Entity;

  /**
   * Game Controller
   */
  export class GameController {

    systems:Systems;

    start() {

      this.systems = this.createSystems(Pools.pool);
      this.systems.initialize();

    }

    createSystems(pool) {
      return new Systems()
        // .add(pool.createSystem(schmupwarz.BackgroundSystem))
        .add(pool.createSystem(schmupwarz.MovementSystem))
        .add(pool.createSystem(schmupwarz.PlayerInputSystem))
        .add(pool.createSystem(schmupwarz.SoundEffectSystem))
        .add(pool.createSystem(schmupwarz.CollisionSystem))
        .add(pool.createSystem(schmupwarz.ExpiringSystem))
        .add(pool.createSystem(schmupwarz.EntitySpawningTimerSystem))
        .add(pool.createSystem(schmupwarz.ParallaxStarRepeatingSystem))
        .add(pool.createSystem(schmupwarz.ColorAnimationSystem))
        .add(pool.createSystem(schmupwarz.ScaleAnimationSystem))
        .add(pool.createSystem(schmupwarz.RemoveOffscreenShipsSystem))
        .add(pool.createSystem(schmupwarz.SpriteRenderSystem))
        .add(pool.createSystem(schmupwarz.AddViewSystem))
        .add(pool.createSystem(schmupwarz.HealthRenderSystem))
        .add(pool.createSystem(schmupwarz.HudRenderSystem))
        .add(pool.createSystem(schmupwarz.DestroySystem));

    }

    update(delta:number) {
      this.systems.execute();
    }
    
    /**
     * Stop the entitas engine
     */
    stop() {
      var entities = Pools.pool.getEntities();
      for (var i = 0, entitiesLength = entities.length; i < entitiesLength; i++) {
        var e:Entity = entities[i];
        e.setDestroy(true);
      }
      /**
      * Reset entitas memory
      * todo: is there a better way to do this?
      */
      Pools._pool = null;
      Entity.alloc = null;    
    }
  }
}
